kv: simplify next() and truncate()

tbg opened the pull request:
The old code was correct, but this is slightly more efficient and
also gives the reader less to think about.

Release notes: none.

Closes #18395.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20103)
<!-- Reviewable:end -->

nvanbenschoten commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/batch.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20103#-Kz5MNytRavZdPkPvOC-:-Kz5MNytRavZdPkPvOC0:b351q58) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7193ca80e06d162bb9f0ad0c68785e1f0bdf04e/pkg/kv/batch.go#L63)):*
> ```Go
> 			return false, emptySpan, err
> 		}
> 		if l, r := keys.IsLocal(header.Key), keys.IsLocal(header.EndKey); l || r {
> ```

Nice :+1: I've seen the previous pattern used a few times in different areas and it always stops me in my tracks for a minute or two.

---

*[pkg/kv/batch.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20103#-Kz5OC_o2uVEt0G0hqJC:-Kz5OC_o2uVEt0G0hqJD:bnal94b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7193ca80e06d162bb9f0ad0c68785e1f0bdf04e/pkg/kv/batch.go#L223)):*
> ```Go
> 		if addr.Less(k) {
> 			if len(h.EndKey) == 0 {
> 				// Affects only [KeyMin,k).
> ```

Could you change here and below to `// h affects only [KeyMin,k), which is less than k.`? Like this, it's easier to read wrong (e.g. "this branch affects only ...").

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20103)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/batch.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20103#-Kz5OC_o2uVEt0G0hqJC:-Kz5RktdvLKKhwaU_xQu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7193ca80e06d162bb9f0ad0c68785e1f0bdf04e/pkg/kv/batch.go#L223)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you change here and below to `// h affects only [KeyMin,k), which is less than k.`? Like this, it's easier to read wrong (e.g. "this branch affects only ...").
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20103)*
<!-- Sent from Reviewable.io -->

tbg committed:
kv: simplify next() and truncate()

The old code was correct, but this is slightly more efficient and
also gives the reader less to think about.

Release notes: none.

Closes #18395.

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20103#-:-Kz5UWwcVtkkNJJZnuhV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!
